platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,175 @@
+- Add labels to the snapshot API
+- The output of the snapshot API should contain the labels
+- The output of the snapshot API must be compatible with the current output (no breaking changes)
+
+## Non-Goals
+
+Even with the new concept the metrics api should be only used in our (hashgraph) code.
+The metrics module should not be used by the end user.
+Here the preferred solution is still that we forward our metrics to a monitoring system that the custom application code uses.
+
+## Design & Architecture
+
+This chapter describes the changes that are needed to add labels to the metrics module.
+
+### Public API
+
+The metrics module will be extended by a new data type called `Label`.
+A label is a key-value pair that can be defined as a `record`:
+
+```java
+record Label(@NonNull String key, @Nullable String value) {}
+``` 
+
+The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:
+
+```java
+interface Metric {
+    
+    //...
+    
+    @NonNull
+    List<Label> getLabels();
+}
+```
+
+> [!IMPORTANT]  
+> I currently ask myself if the method should return `Set`, `Collection` or `List`. Help is appreciated.
+
+Like the name or category, the labels are defined when a metric is created and cannot be changed later.
+Based on that the MetricConfig class will be extended by a new method called `withLabels`:
+
+```java
+abstract class MetricConfig {
+    
+    //...
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Label... labels) {...}
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Collection<Label> labels) {...}
+}
+```
+
+By adding labels the unique identifier of a metric is the key and the labels.
+Based on that we need to modify the `Metrics` interface to use the labels as the unique identifier of a metric:
+
+```java
+
+interface Metrics {
+    
+    // old methods:
+    // Object getValue(@NonNull String category, @NonNull String name);
+    // void remove(@NonNull String category, @NonNull String name);
+    
+    @Nullable
+    Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+    
+    void remove(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+}
+```
+
+Since the old methods are only used in tests or demos today we can do that change easily.
+
+#### Snapshot API
+
+The snapshot API will be extended to return the labels of a metric for a snapshot.
+By doing so any monitoring system that uses the snapshot API can use the labels.
+Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
+Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
+of the snapshot API to the public API.
+
+### Migration
+
+Next to extending the public API we need to migrate the current metrics to the new concept.
+
+#### Migration of the `category`, `name`, and `nodeId`
```

Comment:
We want to add NodeId as a label instead of a value, and I agree with the need there.
But I need to understand why names and categories should be handled as labels. This is not evident in the proposal.

As I said before, there is a difference in that the labels should be optional. You can track values without labels, but you can't without name and category.

On the other hand, one existing problem is tracking the metrics exposed in Grafana to the actual code that modifies their value.  If we complicate the mapping of a metric's name with the code that interacts with it ,